storage: GenericEphemeralVolume (#22438)
kubernetes/website: content/en/docs/concepts/storage/ephemeral-volumes.md

---
reviewers:
- jsafrane
- saad-ali
- msau42
- xing-yang
- pohly
title: Ephemeral Volumes
content_type: concept

> weight: 20

Reviewer: This weight interposes Ephemeral Volumes between the Volumes concept page and the Persistent Volumes concept page. I don't think that's what you want here.

Author: Cut-and-paste without thinking about what this field means... I'm torn between placing it directly after "persistent volumes" and adding it at the end, after all the concepts linked to in the document itself have been introduced.

Let me add it at the end.

---

<!-- overview -->

> This document describes the current state of _ephemeral volumes_ in Kubernetes. Familiarity with [volumes](/docs/concepts/storage/volumes/) is suggested.

Reviewer: Maybe also suggest familiarity with PersistentVolumeClaim and PersistentVolume? Certainly these are relevant.

<!-- body -->

> ## Introduction

Reviewer: I'd omit this heading.

Some application need additional storage but don't care whether that
data is stored persistently across restarts. For example, caching
services are often limited by memory size and can move infrequently
used data into storage that is slower than memory with little impact
on overall performance.

Other applications expect some read-only input data to be present in
files, like configuration data or secret keys.

_Ephemeral volumes_ are designed for these use cases. Because volumes
are created anew for each pod, pods can be stopped and restarted
without being limited to where some persistent volume is available.

Ephemeral volumes are specified _inline_ in the pod spec, which
simplifies application deployment and management.

> 

Reviewer: If you wan, add a heading here: “Types of ephemeral volumes”

Kubernetes supports several different kinds of ephemeral volumes for
different purposes:

> - [emptyDir]((/docs/concepts/volumes/#emptydir): a directory on the root disk or

Reviewer: Does the backing storage `emptyDir` have to live on the node's root filesystem? (I don't believe so).

I'd explain it as “an amount of node-local ephemeral storage, empty at Pod startup”

Reviewer: > The kubelet has a base directory for other locally stored data (`/var/lib/kubelet` by default).
> 
> Typically, `/var/lib/kubelet` is on the system root filesystem, and the kubelet is designed with that layout in mind.

(adapted from https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage)

I don't think “root disk” is correct. First because the root filesystem isn't always backed by durable storage, and second because `/var/lib/kubelet` doesn't have to live on that filesystem.

Author: "root disk" is a simplification. How about "with storage provided locally from the kubelet base directory (usually the root disk) or from memory"?

I'm also not sure about "provided locally from" - perhaps "coming locally from"?

Author: I've switched to "with storage coming locally from the kubelet base directory (usually the root disk) or RAM"

  a tmpfs
- [configMap](/docs/concepts/volumes/#configmap),
  [downwardAPI](/docs/concepts/volumes/#downwardapi),
  [secret](/docs/concepts/storage/volumes/#secret): inject different
  kinds of Kubernetes data into a pod
- [CSI ephemeral
  volumes](docs/concepts/storage/volumes/#csi-ephemeral-volumes):
  similar to the previous volume kinds, but provided by special [CSI
  drivers](https://github.com/container-storage-interface/spec/blob/master/spec.md)
  which specifically [support this feature](https://kubernetes-csi.github.io/docs/drivers.html)

> - _generic ephemeral volumes_ (described [below](#generic-ephemeral-volumes)):
>   can be provided by all storage drivers that also support persistent volumes

Reviewer: nit:
```suggestion
- [generic ephemeral volumes](#generic-ephemeral-volumes), which
  can be provided by all storage drivers that also support persistent volumes
```

- Volumes can have a fixed size that pods are not able to exceed.
- Volumes may have some initial data, depending on the driver and
  parameters.
- All of the normal volume operations
  ([snapshotting](/docs/concepts/storage/volume-snapshots/),
  [cloning](/docs/concepts/storage/volume-pvc-datasource/),
  [resizing](/docs/concepts/storage/persistent-volumes/#expanding-persistent-volumes-claims),
  [storage capacity tracking](/docs/concepts/storage-capacity/), etc.)
  are supported.

Example:

```yaml
kind: Pod
apiVersion: v1
metadata:
  name: my-app
spec:
  containers:
    - name: my-frontend
      image: busybox
      volumeMounts:
      - mountPath: "/scratch"
        name: scratch-volume
      command: [ "sleep", "1000000" ]
  volumes:
    - name: scratch-volume
      ephemeral:
        volumeClaimTemplate:
          metadata:
            labels:
              type: my-frontend-volume
          spec:
            accessModes: [ "ReadWriteOnce" ]
            storageClassName: "scratch-storage-class"
            resources:
              requests:
                storage: 1Gi
```

### Lifecycle and PersistentVolumeClaim

The key design idea is that the [parameters for a

> PersistentVolumeClaim](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#ephemeralvolumesource-v1alpha1-core)

Author: Yes. That "PersistentVolumeClaim" has "persistent" in it is now mostly historic: "VolumeClaim" would have been a more appropriate name for the API. We can't fix the API, but in prose we can just refer to it as "volume claim".

I changed the link text but then kept `PersistentVolumeClaim object` in the rest of the sentence because technically that's what it is.

are allowed inside a volume source of the pod. Labels, annotations and
the full PersistentVolumeClaimSpec are supported. When such a pod gets
created, a new controller then creates an actual PersistentVolumeClaim

> object in the same namespace as the pod.

Author: "new for 1.19" - replaced with "the ephemeral volume controller" because that's how it is identified in log messages and source code.

That triggers volume binding and/or provisioning, either immediately if
the storage class uses immediate volume binding or when the pod is
tentatively scheduled onto a node (`WaitForFirstConsumer` volume
binding mode). The latter is recommended for generic ephemeral volumes
because then the pod scheduler is free to choose a suitable node for
the pod. With immediate binding, it is forced to use a node that has
access to the volume once it is available.

> These additional PVCs are owned by the pod. When the pod gets deleted,

Author: Makes sense.

the Kubernetes garbage collector deletes the PVC, which then usually
triggers deletion of the volume because the default reclaim policy of
storage classes is to delete volumes. If for some reason an ephemeral
volume is not meant to be deleted, a storage class with "retain" as

> reclaim policy can be used.

Author: Looks good.

Once these PVCs exist, they can be used like any other PVC. In
particular, they can be referenced as data source in volume cloning or
snapshotting. The PVC object also holds the current status of the
volume.

### PVC Naming

Naming of the additional PVCs is currently deterministic: the name is
a combination of pod name and volume name, with a hyphen (`-`) in the
middle. In the example above, the PVC name will be
`my-app-scratch-volume`.  This deterministic naming makes it easier to
interact with the PVC because one does not have to search for it once
the pod name and volume name are known.

> However, it also introduces a potential conflict between different

Author: Changed.

volume if it was created for the pod. This check is based on the
ownership relationship. An existing PVC is not overwritten or
modified. But this does not resolve the conflict because without the
right PVC, the pod cannot start.

Therefore care must be taken when naming pods and volumes inside the
same namespace such that these conflicts cannot occur.

### Security

Enabling the GenericEphemeralVolume feature allows users to create
PVCs indirectly if they can create pods, even if they do not have
permission to create them directly. Cluster administrators must be
aware of this. If this does not fit their security model, they have
two choices:
- Explicitly disable the feature through the feature gate, to avoid
  being surprised when some future Kubernetes version enables it
  by default.
- Use a [Pod Security
  Policy](/docs/concepts/policy/pod-security-policy/) where the
  `volumes` list does not contain the `ephemeral` volume type.

The normal namespace quota for PVCs in a namespace still applies, so
even if users are allowed to use this new mechanism, they cannot use
it to circumvent other policies.

## {{% heading "whatsnext" %}}

> 

Reviewer: Please also signpost readers to https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage

Author: Added:
```
### Ephemeral volumes managed by kubelet

See [local ephemeral storage](/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage).
```

---
reviewers:
- jsafrane
- saad-ali
- msau42
- xing-yang
- pohly
title: Ephemeral Volumes
content_type: concept
weight: 50
---

<!-- overview -->

This document describes _ephemeral volumes_ in Kubernetes. Familiarity
with [volumes](/docs/concepts/storage/volumes/) is suggested, in
particular `PersistentVolumeClaim` and `PersistentVolume`.

<!-- body -->

Some application need additional storage but don't care whether that
data is stored persistently across restarts. For example, caching
services are often limited by memory size and can move infrequently
used data into storage that is slower than memory with little impact
on overall performance.

Other applications expect some read-only input data to be present in
files, like configuration data or secret keys.

> _Ephemeral volumes_ are designed for these use cases. Because volumes

Reviewer 2: I think it's important to mention somehwere in the intro that ephemeral volumes follow the pod's lifetime and get created and deleted along with the pod.

Author: Replaced "created anew for each pod" with "follow the pod's lifetime and get created and deleted along with the pod".

Or rather, *Pod* (capital case)... I also fixed that elsewhere.

Kubernetes supports several different kinds of ephemeral volumes for
different purposes:
- [emptyDir]((/docs/concepts/volumes/#emptydir): empty at Pod startup,
  with storage provided locally from the root disk or RAM
- [configMap](/docs/concepts/volumes/#configmap),
  [downwardAPI](/docs/concepts/volumes/#downwardapi),
  [secret](/docs/concepts/storage/volumes/#secret): inject different
  kinds of Kubernetes data into a pod
- [CSI ephemeral
  volumes](docs/concepts/storage/volumes/#csi-ephemeral-volumes):
  similar to the previous volume kinds, but provided by special [CSI
  drivers](https://github.com/container-storage-interface/spec/blob/master/spec.md)
  which specifically [support this feature](https://kubernetes-csi.github.io/docs/drivers.html)
- [generic ephemeral volumes](#generic-ephemeral-volumes), which
  can be provided by all storage drivers that also support persistent volumes

`emptyDir`, `configMap`, `downwardAPI`, `secret` are provided as
[local ephemeral
storage](/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage).
They are managed by kubelet on each node.

CSI ephemeral volumes *must* be provided by third-party CSI storage

> drivers. Generic ephemeral volumes *can* be provided by third-party

Reviewer 2: Start "Generic ephemeral volumes on a separate line" since we're talking about a different feature.

Author: Okay.

> CSI storage drivers, but also by any other storage driver that

Reviewer 2: The same storage drivers that support CSI ephemeral volumes may not be necessarily able to support generic ephemeral volumes.

Author: Added: "Some CSI drivers are written specifically for CSI
ephemeral volumes and do not support dynamic provisioning: those then
cannot be used for generic ephemeral volumes."

> supports dynamic provisioning. These drivers can offer functionality

Author: "These drivers" became a bit ambiguous. I replaced it with "The advantage of using third-party drivers is that they can..."

      csi:
        driver: inline.storage.kubernetes.io
        volumeAttributes:
          foo: bar
```

The `volumeAttributes` determine what volume is prepared by the
driver. These attributes are specific to each driver and not
standardized. See the documentation of each CSI driver for further
instructions.

As a cluster administrator, you can use a [PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/) to control which CSI drivers can be used in a Pod, specified with the
[`allowedCSIDrivers` field](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#podsecuritypolicyspec-v1beta1-policy).

### Generic ephemeral volumes

{{< feature-state for_k8s_version="v1.19" state="alpha" >}}

This feature requires the `GenericEphemeralVolume` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) to be
enabled. Because this is an alpha feature, it is disabled by default.

Generic ephemeral volumes are similar to `emptyDir` volumes, just more
flexible:
- Storage can be local or network-attached.
- Volumes can have a fixed size that pods are not able to exceed.
- Volumes may have some initial data, depending on the driver and
  parameters.

> - Typical operations on volumes are supported, including

Reviewer 2: assuming that the driver supports them

Author: Added.

  and [storage capacity tracking](/docs/concepts/storage-capacity/).

Example:

```yaml
kind: Pod
apiVersion: v1
metadata:
  name: my-app
spec:
  containers:
    - name: my-frontend
      image: busybox
      volumeMounts:
      - mountPath: "/scratch"
        name: scratch-volume
      command: [ "sleep", "1000000" ]
  volumes:
    - name: scratch-volume
      ephemeral:
        volumeClaimTemplate:
          metadata:
            labels:
              type: my-frontend-volume
          spec:
            accessModes: [ "ReadWriteOnce" ]
            storageClassName: "scratch-storage-class"
            resources:
              requests:
                storage: 1Gi
```

### Lifecycle and PersistentVolumeClaim

The key design idea is that the [parameters for a
volume claim](/docs/reference/generated/kubernetes-api/#ephemeralvolumesource-v1alpha1-core)
are allowed inside a volume source of the pod. Labels, annotations and
the whole set of fields for a PersistentVolumeClaim are supported. When such a Pod gets
created, the ephemeral volume controller then creates an actual PersistentVolumeClaim

> object in the same namespace as the pod.

Reviewer 2: And gets deleted when the Pod is deleted.

Author: I've added "... and ensures that the PersistentVolumeClaim gets deleted when the Pod gets deleted." How that works specifically is then explained later - I didn't want to repeat that explanation.

IMHO explaining the steps in chronological order makes sense.

That triggers volume binding and/or provisioning, either immediately if
the {{< glossary_tooltip text="StorageClass" term_id="storage-class" >}} uses immediate volume binding or when the pod is
tentatively scheduled onto a node (`WaitForFirstConsumer` volume
binding mode). The latter is recommended for generic ephemeral volumes
because then the scheduler is free to choose a suitable node for
the Pod. With immediate binding, the scheduler is forced to select a node that has
access to the volume once it is available.

In terms of [resource ownership](/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents),
a Pod that has generic ephemeral storage is the owner of the PersistentVolumeClaim(s)
that provide that ephemeral storage. When the Pod is deleted,
the Kubernetes garbage collector deletes the PVC, which then usually
triggers deletion of the volume because the default reclaim policy of
storage classes is to delete volumes. You can create quasi-ephemeral local storage
using a StorageClass with a reclaim policy of `retain`: the storage outlives the Pod, 
and in this case you need to ensure that volume clean up happens separately.

> Once these PVCs exist, they can be used like any other PVC. In

Author: Replaced "Once" with "While" - it's more accurate and hints that this can change...

particular, they can be referenced as data source in volume cloning or
snapshotting. The PVC object also holds the current status of the
volume.

### PersistentVolumeClaim naming

> Naming of the additional PVCs is deterministic: the name is

Reviewer 2: additional -> ephemeral?  although "ephemeral PVC" is a bit of an oxymoron

Author: Yes, that's why I avoided it :cry: 

Replaced with "automatically created".

a combination of pod name and volume name, with a hyphen (`-`) in the
middle. In the example above, the PVC name will be
`my-app-scratch-volume`.  This deterministic naming makes it easier to
interact with the PVC because one does not have to search for it once
the pod name and volume name are known.

The deterministic naming also introduces a potential conflict between different
pods (a pod "pod-a" with volume "scratch" and another pod with name
"pod" and volume "a-scratch" both end up with the same PVC name
"pod-a-scratch") and between pods and manually created PVCs.

Such conflicts are detected: a PVC is only used for an ephemeral
volume if it was created for the pod. This check is based on the
ownership relationship. An existing PVC is not overwritten or
modified. But this does not resolve the conflict because without the
right PVC, the pod cannot start.

{{< caution >}}
Take care when naming Pods and volumes inside the
same namespace, so that these conflicts can't occur.
{{< /caution >}}

### Security

Enabling the GenericEphemeralVolume feature allows users to create
PVCs indirectly if they can create pods, even if they do not have

> permission to create them directly. Cluster administrators must be

Reviewer 2: them -> PVCs

Author: Changed.

---
reviewers:
- jsafrane
- saad-ali
- msau42
- xing-yang
- pohly
title: Ephemeral Volumes
content_type: concept
weight: 50
---

<!-- overview -->

This document describes _ephemeral volumes_ in Kubernetes. Familiarity
with [volumes](/docs/concepts/storage/volumes/) is suggested, in

> particular `PersistentVolumeClaim` and `PersistentVolume`.

Reviewer: nit:
```suggestion
particular PersistentVolumeClaim and PersistentVolume.
```

<!-- body -->

Some application need additional storage but don't care whether that
data is stored persistently across restarts. For example, caching
services are often limited by memory size and can move infrequently
used data into storage that is slower than memory with little impact
on overall performance.

Other applications expect some read-only input data to be present in
files, like configuration data or secret keys.

_Ephemeral volumes_ are designed for these use cases. Because volumes
follow the Pod's lifetime and get created and deleted along with the
Pod, Pods can be stopped and restarted without being limited to where
some persistent volume is available.

Ephemeral volumes are specified _inline_ in the Pod spec, which
simplifies application deployment and management.

### Types of ephemeral volumes

Kubernetes supports several different kinds of ephemeral volumes for
different purposes:

> - [emptyDir]((/docs/concepts/volumes/#emptydir): empty at Pod startup,

Reviewer: Invalid Markdown here I'm afraid
```suggestion
- [emptyDir](/docs/concepts/volumes/#emptydir): empty at Pod startup,
```

  by default.
- Use a [Pod Security
  Policy](/docs/concepts/policy/pod-security-policy/) where the
  `volumes` list does not contain the `ephemeral` volume type.

The normal namespace quota for PVCs in a namespace still applies, so
even if users are allowed to use this new mechanism, they cannot use
it to circumvent other policies.

## {{% heading "whatsnext" %}}

### Ephemeral volumes managed by kubelet

See [local ephemeral storage](/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage).

### CSI ephemeral volumes

- For more information on the design, see the [Ephemeral Inline CSI
  volumes KEP](https://github.com/kubernetes/enhancements/blob/ad6021b3d61a49040a3f835e12c8bb5424db2bbb/keps/sig-storage/20190122-csi-inline-volumes.md).
- For more information on further development of this feature, see the [enhancement tracking issue #596](https://github.com/kubernetes/enhancements/issues/596).

### Generic ephemeral volumes

- For more information on the design, see the
[Generic ephemeral inline volumes KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1698-generic-ephemeral-volumes/README.md).

> - For more information on further development of this feature, see the [enhancement tracking issue #1698](https://github.com/kubernetes/enhancements/issues/1698).

Reviewer: This could be a little tidier (for example, we usually and intentionally don't link to KEPs other than for historical context), but I'm happy to log a separate issue to track tidying that up after this merges. It's a minor detail for sure.

Author: I think that for alpha and beta features, easy access to both the KEP and the enhancement issue are important: the KEP explains "why" the feature is what it is and in particular, what other alternatives were considered. The enhancement issue answers the questions "when will this graduate" and "what is still changing with this feature".

I can remove it in a follow-up PR if this explanation is not convincing enough. But for now, let's merge.
